python/lsst/meas/astrom/matchPessimisticB.py

```diff
@@ -472,23 +476,82 @@ def _doMatch(self, refCat, sourceCat, wcs, refFluxField, numUsableSources,
                 match_tolerance.lastMatchedPattern = matcher_struct.pattern_idx
                 break
 
+        # The matcher returns all the nearst neighbors between the reference
+        # source catalogs. For the currrent astrometry solver we need to clip
+        # these down to give the best solution.
+        distances = np.degrees(matcher_struct.distances) * 3600
+        dist_cut = np.max(
+            (2 * wcs.pixelScale().asArcseconds(),
+             self._sigma_clip(
+                distances, 2,
+                np.max((10 * wcs.pixelScale().asArcseconds(),
+                        maxMatchDistArcSec,
+                        maxShiftArcseconds)))))
+
         # A match has been found, return our list of matches and
         # return.
         matches = []
         for match_ids, dist in zip(matcher_struct.matches,
-                                   matcher_struct.distances):
-            match = afwTable.ReferenceMatch()
-            match.first = refCat[match_ids[1]]
-            match.second = sourceCat[match_ids[0]]
-            match.distance = match.first.getCoord().angularSeparation(
-                match.second.getCoord())
-            matches.append(match)
+                                   distances):
+            if dist < dist_cut:
+                match = afwTable.ReferenceMatch()
+                match.first = refCat[match_ids[1]]
+                match.second = sourceCat[match_ids[0]]
+                match.distance = match.first.getCoord().angularSeparation(
+                    match.second.getCoord())
```

Comment:
Why can't you use the distance you already know? If you can't, then a code comment explaining would be helpful.

Also I find it very confusing to have `matches` use our usual definition of a list of `ReferenceMatch` here, but `matcher_struct.matches` mean something entirely different. For now I suggest renaming `matcher_struct.matches` to something else, such as `matched_ids`. But I hope in the long run `matcher_struct` can contain a normal match list (list of `ReferenceMatch`) and get rid of the separate distance field.